.github/workflows/ubuntu_21_04.yml

```diff
@@ -6,7 +6,10 @@ on:
   repository_dispatch:
     types: [backend_automation]
   workflow_dispatch:
-
+concurrency:
+  group: ${{ github.workflow }}-${{ github.ref }}
```

Comment:
Done.

Reply:
> Any previously pending job or workflow in the concurrency group will be canceled.

So we may miss testing for a previous push to `master`.

We should take care to tags.

It is better to check for 1.10 directly, not using startWith().

Nope, github.sha is not enough (see comment below).

I guess the following should work (not tested):

```yaml
# Updating of a developer branch cancels previously scheduled runs
# of this branch. However `master`, release branch (`1.10`,
# `2.10`, ...) and tag workflow runs are never cancelled.
# 
# We use a trick here: define the concurrency group as 'workflow
# run id' + 'workflow run attempt', because it is unique for any
# run attempt. So it effectively discards grouping.
# 
# Important: we cannot use `github.sha` as an unique identifier,
# because pushing a tag may cancel a run that works on a branch
# pushing event. I guess we can use ref + sha, but I'm not
# completely sure that it covers all cases. So I prefer 'run id' +
# 'run attempt', whose uniqueness seems to be guaranteed by
# GitHub.
concurrency:
  group: >
    ${{ (
      startsWith(github.ref, refs/tags/') ||
      github.ref == 'refs/heads/master' ||
      github.ref == 'refs/heads/1.10' ||
      startsWith(github.ref, 'refs/heads/2.')) &&
    format('run_id: {0}; run_attempt: {1}', github.run_id, github.run_attempt) ||
    format('workflow: {0}; ref: {1}', github.workflow. github.ref) }}
  cancel-in-progress: true
```

It uses the fake ternary operator (`<cond> && value_1 || value_2`), folks report that it works on GitHub:

https://github.com/actions/runner/issues/409
https://github.com/nektos/act/issues/718

Reply:
Your idea about the fake ternary operator is awesome! In this case, we don't need a separate action to get the concurrency group. Thanks!

Reply:
Well, I have nothing more to say there: I see no problems anymore.